query-engine/query-engine/src/server/mod.rs

```diff
@@ -0,0 +1,199 @@
+pub struct HttpServerBuilder {
+    /// The address we listen on.
+    addr: SocketAddr,
+    /// The server configuration passed.
+    config: Configuration,
+    /// The Prisma data model.
+    datamodel: Datamodel,
+    /// Are we listening in legacy mode?
+    legacy_mode: bool,
+    /// Do we enable raw queries?
+    ///
+    /// Note: this has security implications.
+    enable_raw_queries: bool,
+    /// Do we enable the GraphQL playground?
+    ///
+    /// Note: this has security implications.
+    enable_playground: bool,
+}
+
+impl HttpServerBuilder {
+    /// Create a new instance of `HttpServerBuilder`.
+    fn new(addr: SocketAddr, config: Configuration, datamodel: Datamodel) -> Self {
+        Self {
+            addr,
+            config,
+            datamodel,
+            legacy_mode: false,
+            enable_playground: false,
+            enable_raw_queries: false,
+        }
+    }
+
+    /// Enable "legacy mode" for prisma-engines.
+    pub fn legacy(mut self, val: bool) -> Self {
+        self.legacy_mode = val;
+        self
+    }
+
+    /// Enable raw queries for prisma-engines.
+    ///
+    /// # Security
+    ///
+    /// Enabling this setting will allow arbtrary queries to be executed against
+    /// the server. This has security implications when exposing Prisma on a
+    /// public port.
+    pub fn enable_raw_queries(mut self, val: bool) -> Self {
+        self.enable_raw_queries = val;
+        self
+    }
+
+    /// Enable the GraphQL playground.
+    pub fn enable_playground(mut self, val: bool) -> Self {
+        self.enable_playground = val;
+        self
+    }
+
+    /// Start the server.
+    pub async fn build(self) -> PrismaResult<()> {
+        let ctx = PrismaContext::builder(self.config, self.datamodel)
+            .legacy(self.legacy_mode)
+            .enable_raw_queries(self.enable_raw_queries)
+            .build()
+            .await?;
+
+        HttpServer::run(self.addr, ctx, self.enable_playground).await
+    }
+}
+
+pub struct HttpServer;
+
+impl HttpServer {
+    /// Create a new HTTP server builder.
+    pub fn builder(addr: SocketAddr, config: Configuration, datamodel: Datamodel) -> HttpServerBuilder {
+        HttpServerBuilder::new(addr, config, datamodel)
+    }
+
+    /// Start the HTTP server with the default options enabled.
+    async fn run(addr: SocketAddr, cx: PrismaContext, enable_playground: bool) -> PrismaResult<()> {
+        let mut app = tide::with_state(State::new(cx, enable_playground));
+        app.middleware(ElapsedMiddleware::new());
+
+        app.at("/").post(graphql_handler);
+        app.at("/").get(playground_handler);
+        app.at("/sdl").get(sdl_handler);
+        app.at("/dmmf").get(dmmf_handler);
+        app.at("/server_info").get(server_info_handler);
+        app.at("/status").get(|_| async move { Ok(json!({"status": "ok"})) });
+
+        info!("Started http server on {}:{}", addr.ip(), addr.port());
+        app.listen(addr).await?;
```

Comment:
Switching to Unix streams is a matter of replacing this line with:
```rust
app.listen_unix(addr).await?;
```

Reply:
And I guess on windows:
```
#[cfg(windows)]
app.listen_windows(addr).await?;
```